### PR TITLE
Alire.Config: add support of built-in options

### PR DESCRIPTION
**Pull request by Fabien-Chouteau**
Built-in options have a type that is checked when setting or loading a configuration file.
They also have a short help text to explain their meaning and usage.

### CODE REVIEW

File: doc/configuration.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    otherwise the command fails.
+
+
+## Custom configuration options
+
+The `alr config` command allows you to set an get any combination of
+configuration option `key` and `value`. You can use this feature to store you
+own project related configuration, or implement tools that integrate in an
+`Alire` context. However, be carful when naming custom configuration options as
+`Alire` my use the same `key` in the future. We recomend using a distinctive
```
[the remark this answers is not in the document]
Comment (Fabien-Chouteau) [reply doc/configuration.md]: Well, what a pleasant surprise. Thank you misspell! :)

File: doc/configuration.md [hunk 2]
```diff
@@ -41,9 +41,9 @@
 The `alr config` command allows you to set an get any combination of
 configuration option `key` and `value`. You can use this feature to store you
 own project related configuration, or implement tools that integrate in an
-`Alire` context. However, be carful when naming custom configuration options as
-`Alire` my use the same `key` in the future. We recomend using a distinctive
-sub-categorie name, for instance: `my_project.my_config_option`.
+`Alire` context. However, be careful when naming custom configuration options as
+`Alire` my use the same `key` in the future. We recommend using a distinctive
```
Comment (mosteo) [inline doc/configuration.md]: my -> may

**Review by mosteo — APPROVED** [review]

**Comment by mosteo** [comment]
Great addition, Fabien.

Not from this PR, but I just noticed it: in `alire-config.adb:280` the file name is `config.toml`, and in line 289 it's `config` (no extension). Is this on purpose?

**Comment by Fabien-Chouteau** [comment]
>  the file name is `config.toml`, and in line 289 it's `config` (no extension). Is this on purpose?

It is on purpose, but it is not clean. I didn't think to ask for feedback on this point.

I wanted to use `config.toml` everywhere, but adding another `.toml` file in the `alire` directory broke everything. It think it has to do with the way crate context is detected. Basically `Alire` was seeing a `config` crate.

**Comment by mosteo** [comment]
> I wanted to use `config.toml` everywhere, but adding another `.toml` file in the `alire` directory broke everything. It think it has to do with the way crate context is detected. Basically `Alire` was seeing a `config` crate.

Ah, right. We use the name of the only '.toml' file as the crate name. This is something that has bothered me for a while but I'm not sure if it's worth the pain to change it.

What do you think of using `alire.conf`/`alire.config` instead? Although as it is, it's OK for me.

**Comment by mosteo** [comment]
Merging already because the file naming does not belong to this PR anyway. Thanks!

**Comment by Fabien-Chouteau** [comment]
> What do you think of using `alire.conf`/`alire.config` instead? Although as it is, it's OK for me.

Maybe just `config` like git?

**Comment by mosteo** [comment]
Fine too, sure

**Pull request merged**